java-checks/src/main/java/org/sonar/java/checks/FinalizeFieldsSetCheck.java

```diff
@@ -0,0 +1,114 @@
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02
+ */
+package org.sonar.java.checks;
+
+import com.google.common.collect.ImmutableList;
+import org.apache.commons.lang.BooleanUtils;
+import org.sonar.check.BelongsToProfile;
+import org.sonar.check.Priority;
+import org.sonar.check.Rule;
+import org.sonar.java.model.declaration.MethodTreeImpl;
+import org.sonar.java.resolve.Symbol;
+import org.sonar.plugins.java.api.tree.AssignmentExpressionTree;
+import org.sonar.plugins.java.api.tree.BaseTreeVisitor;
+import org.sonar.plugins.java.api.tree.ClassTree;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.MemberSelectExpressionTree;
+import org.sonar.plugins.java.api.tree.MethodTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.Tree.Kind;
+
+import java.util.List;
+
+@Rule(
+  key = "S2165",
+  priority = Priority.MAJOR,
+  tags = {"clumsy", "performance"})
+@BelongsToProfile(title = "Sonar way", priority = Priority.MAJOR)
+public class FinalizeFieldsSetCheck extends SubscriptionBaseVisitor {
+
+  @Override
+  public List<Kind> nodesToVisit() {
+    return ImmutableList.of(Kind.METHOD);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    MethodTree methodTree = (MethodTree) tree;
+    if (isFinalizeDeclaration(methodTree)) {
+      methodTree.accept(new AssignmentVisitor());
+    }
+  }
+
+  private boolean isFinalizeDeclaration(MethodTree tree) {
+    return hasSemantic() && isMethodNamedFinalize(tree) && isOverriding(tree) && hasNoParameters(tree);
+  }
+
+  private boolean isMethodNamedFinalize(MethodTree tree) {
+    return "finalize".equals(tree.simpleName().name());
+  }
+
+  private boolean isOverriding(MethodTree tree) {
+    return BooleanUtils.isTrue(((MethodTreeImpl) tree).isOverriding());
+  }
+
+  private boolean hasNoParameters(MethodTree tree) {
+    return tree.parameters().isEmpty();
+  }
+
+  private class AssignmentVisitor extends BaseTreeVisitor {
+    @Override
+    public void visitClass(ClassTree tree) {
+      // Do not visit inner classes as their methods will be visited by main visitor
+    }
+
+    @Override
+    public void visitAssignmentExpression(AssignmentExpressionTree tree) {
+      if (isFieldAssignment(tree) && isNullAssignment(tree)) {
+        addIssue(tree, "Remove this nullification of \"" + getFieldName(tree) + "\".");
+      }
+    }
+
+    private boolean isFieldAssignment(AssignmentExpressionTree tree) {
+      ExpressionTree variable = tree.variable();
+      if (variable.is(Kind.MEMBER_SELECT)) {
+        variable = ((MemberSelectExpressionTree) variable).expression();
+      }
+      if (variable.is(Kind.IDENTIFIER)) {
+        Symbol variableSymbol = getSemanticModel().getReference((IdentifierTree) variable);
+        return variableSymbol.owner().isKind(Symbol.TYP);
```

Comment:
This is not sufficient if you have an expression refering a field of another instance  or from another class : 
class A  {
  private int field;
  protected void finalize() {
    new A().field = null;
    new B().field = null;
  }
}